Rework nRF5 entropy driver

pizi-nordic opened the pull request:
This PR is an enhanced version of the #9593. I hope, that all comments posted there are addressed.
Changes:
- drivers: entropy: nrf5: Use nrf_rng HAL for all hardware access
- drivers: entropy: nrf5: Fix race condition on hardware access
- drivers: entropy: nrf5: Cleanup function naming
- drivers: entropy: nrf5: Pass device data to the ISR
- drivers: entropy: nrf5: Replace random number pool implementation
- drivers: entropy: nrf5: Remove entropy_nrf_get_entropy_isr() function
- drivers: entropy: Use likely()/unlikely() macros to speed up the API
- Bluetooth: controller: Optimize RNG access

Notes for Bluetooth guys:
- Performance:
   - After this PR:   [bt] [INF] encode_control: l: 7, 6, 9; t: 56, 39, 118.
   - Before this PR: [bt] [INF] encode_control: l: 6, 6, 8; t: 53, 46, 122.
- Unfortunately I was not able to develop fast & lockless algorithm allowing for safe shared access to the RNG pool on the Cortex-M0. However I ensured, that interrupts are locked for fixed and small amount of time (few instructions).

pizi-nordic committed:
drivers: entropy: nrf5: Use nrf_rng HAL for all hardware access

The nrf_rng HAL uses barriers on SoCs which need them to ensure
that all register writes will reach the peripheral immediately.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Remove entropy_nrf_get_entropy_isr() function

The entropy_nrf_get_entropy_isr(), which is specific to this driver,
is in fact equivalent of generic entropy_get_entropy_isr(..., 0).

This commit removes the entropy_nrf_get_entropy_isr() function
and replaces its usage by call to generic entropy API.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: Use likely()/unlikely() macros to speed up the API

This commit introduces usage of likely()/unlikely() macros in order
to speed up the entropy_get_entropy_isr() API.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
Bluetooth: controller: Optimize RNG access

This commit optimizes access to RNG driver by taking advantage
of the data structures layout. As result, number of calls to RNG
driver is reduced.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Fix race condition on hardware access

Neither entropy_nrf5_get_entropy_isr() nor isr() protected
access to hardware. As result in some situations a random value
could be "stolen" by a higher priority call  interrupting one
of the mentioned functions.

This commit solves this problem by introducing random_byte_get()
which accesses hardware atomically.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Fix hardware state management

The entropy_nrf5_get_entropy_isr(...,  ENTROPY_BUSYWAIT)
unconditionally started and stopped hardware random number generator
disrupting management of random data pools implemented by the driver.

Moreover, simultaneous call to mentioned function from different
priorities led to hangup of lower priority call as hardware was
stopped by higher priority call while the other one waited for data.

This commit solves mentioned problems by moving the responsibility
for stopping hardware to the interrupt handler. As result hardware
is stopped only when there is no blocking calls to the
entropy_nrf5_get_entropy_isr() as well as all random data pools
are full.

This commit also fixes interrupt management in the mentioned
function, as previous implementation unconditionally enabled
interrupt causing problems similar to the described above.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Cleanup function naming

This commint replaces function names by more accurate ones.
For example, the isr(), was renamed to rng_pool_put(), which
clerarly indicates what this function is doing.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Pass device data to the ISR

This commit changes the argument passed to ISR to the pointer
to device data structure, removing one dereference in the interrupt
handler.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: entropy: nrf5: Replace random number pool implementation

The old algorithm managing random number pools was inefficient and
prone to race conditions. Moreover, its behaviour was not compatible
with the API (provided buffers were filled from the end, leaving
beggining of the buffer untouched if there was not enough random data).

The new implementation is a bit faster and it is race-free as long as
all callers of the entropy_get_entropy_isr() are executed in
"run-to-completion" mode.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,165 @@
 
-	if (remaining < rng->threshold) {
+	len = dst - buf;
+	available = available - len;
```
You could move this line before `while (likely(len--))`. Then this reconstruction of `len` will be not needed.
Actually, a smart compiler should be able to make such optimization, but...

anangl replied:
Right. I forgot about the return value.

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,165 @@
+	ret = rng_pool_put((struct rng_pool *)dev_data->isr, byte);
+	if (ret < 0) {
+		ret = rng_pool_put((struct rng_pool *)dev_data->thr, byte);
+		if (ret) {
```
These two checks of `ret` should be consistent. Now it looks like a positive value may occur and that it is acceptable in the first case and not in the second one.

pizi-nordic replied:
Fixed.

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -272,44 +233,44 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len
-	if (!(flags & ENTROPY_BUSYWAIT)) {
-		return get((struct rand *)dev_data->isr, len, buf);
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		return rng_pool_get((struct rng_pool *)dev_data->isr, buf, len);
```
In this case `len` is not checked against 0 but it is in all other cases (`ENTROPY_BUSYWAIT` and `entropy_nrf5_get_entropy`). It would be good to keep consistency.

pizi-nordic replied:
The `rng_pool_get()` works correctly if len = 0. Here, the check is skipped, as we have to be really fast.

anangl replied:
I propose another optimization then. Since we've got only one entropy device here, there is no need to use `DEV_DATA(dev)` in this function. You could use `entropy_nrf5_data` directly and save a few cycles more.

pizi-nordic replied:
Good point. Optimization added.

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -272,44 +233,44 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len
-			while (NRF_RNG->EVENTS_VALRDY == 0) {
+			int byte;
+
+			while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
```
I don't entirely get this approach. Why for the flag `ENTROPY_BUSYWAIT` the pool is not used first, and the busy loop involved only if there is not enough random data available?
This flag is described in the API as "Busy-wait for random data to be ready". Does it imply that the random data should be generated in the busy loop (to be "fresh", not stored previously for some time)?

pizi-nordic replied:
This approach is simple and works and correctly in the case when request for random data comes from priority higher than RNG interrupt priority. Also, it leaves ISR pool untouched for the services which needs random data *fast*.

anangl replied:
It's probably worth a word of comment in the code then.

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,165 @@
+	first = rngp->first_alloc;
+
+	other_read_in_progress = (rngp->first_read ^ first);
+	available = (last - first) & mask;
```
What about returning here (after unlocking interrupts, of course) when no random data is available? Will this be a rare case, not worth optimizing?

pizi-nordic replied:
It will decrease performance of more common case when random data is available. We have about 20us to handle the random data request from BLE, which is equals to about 320 CPU cycles on nRF51. Most of this time is already taken by API abstraction when we are reaching here.

pizi-nordic commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -277,38 +277,35 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len
 
 	if (len) {
-		u32_t intenset;
+		u32_t valrdy_int_enabled;
```
The mentioned code is removed in one of subsequent commits.

pizi-nordic commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,165 @@
-	u8_t first;
+struct rng_pool {
+	u8_t first_alloc;
+	u8_t first_read;
```
I will add proper description of the algorithm.

pizi-nordic replied:
I added some comments clarifying pool access algorithm.

carlescufi commented:
@pizi-nordic can you please run the following tests on this PR and confirm they all pass:

* tests/kernel/mem_protect/stackprot
* tests/kernel/mem_protect/stack_random
* tests/crypto/rand32. In particular: crypto.rand32.random_hw_xoroshiro.rand32, with prj_hw_random_xoroshiro.conf

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,171 @@
+	u8_t mask  = rngp->mask;
 
-	if (!rng) {
+	/* Signal error if the ool is full. */
```
pool

pizi-nordic replied:
Fixed.

pizi-nordic commented:
>@pizi-nordic can you please run the following tests on this PR and confirm they all pass:
>tests/kernel/mem_protect/stackprot

nRF51: PASS
nRF52: PASS

>tests/kernel/mem_protect/stack_random

nRF51: PASS
nRF52: PASS

>tests/crypto/rand32. In particular: crypto.rand32.random_hw_xoroshiro.rand32, with prj_hw_random_xoroshiro.conf

nRF51: PASS
nRF52: PASS

pizi-nordic committed:
drivers: entropy: nrf5: Save few cycles removing excessive dereferences.

This commit optimizes access to device data using direct reference
instead of relying on the driver_data field in the device structure.

Such approach saves few cycles on time-critical paths.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

anangl commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -269,53 +241,55 @@ static int entropy_nrf5_get_entropy(struct device *device, u8_t *buf, u16_t len)
-	if (!(flags & ENTROPY_BUSYWAIT)) {
-		return get((struct rand *)dev_data->isr, len, buf);
+	/* Check if this API is called on correct driver instance. */
+	__ASSERT_NO_MSG(&entropy_nrf5_data == DEV_DATA(dev));
```
I thought you wanted this to be fast. And now with this assertion it will be even slower than in previous version.

anangl replied:
I'd rather get rid of the `DEV_DATA` macro completely (not use the `driver_data` field from the `device` structure at all) to indicate somehow that this driver handles one instance only.

pizi-nordic replied:
Normally asserts are not compiled in. I would like to keep the asserts in order to check if our assumptions are still correct.

cvinayak reviewed (changes requested):
In general LGTM. ~~But now the nRF51 hard faults not meeting the tIFS when FAST_ENC = n, i.e. without -Ofast, normal re-encryption usecase asserts :-(~~. On further testing these failures exist in upstream/master prior to this PR (enabling data length and PHY update procedure supports in nRF51).

carlescufi commented:
> > @pizi-nordic can you please run the following tests on this PR and confirm they all pass:
> > tests/kernel/mem_protect/stackprot
> 
> nRF51: PASS
> nRF52: PASS
> 
> > tests/kernel/mem_protect/stack_random
> 
> nRF51: PASS
> nRF52: PASS
> 
> > tests/crypto/rand32. In particular: crypto.rand32.random_hw_xoroshiro.rand32, with prj_hw_random_xoroshiro.conf
> 
> nRF51: PASS
> nRF52: PASS

Thank you!

carlescufi commented:
@pizi-nordic can you please rebase?

carlescufi commented:
@cvinayak please re-review
@pizi-nordic CI failing

pizi-nordic commented:
@carlescufi: Whitespace error, fixed.

cvinayak commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -63,204 +63,176 @@
+				     buf, len);
+		k_sem_give(&entropy_nrf5_data.sem_lock);
+
+		if (bytes == 0) {
```
Ok, we can take optimizing towards a power save behavior in future when we observe, if any.

aescolar reviewed (commented):
No complaints from my side

aescolar commented:
@carlescufi : Thanks for including me in the review.
